### PR TITLE
Issue#16 follow up visit schedules

### PR DESCRIPTION
**Pull request by zzainulabidin**

### CODE REVIEW

File: opensrp-fp/src/main/java/org/smartregister/fp/common/interactor/ContactInteractor.java [hunk 1]
```diff
@@ -139,11 +128,11 @@ public void fetchWomanDetails(String baseEntityId, BaseContactContract.Interacto
             } catch (Exception e) {
                 Timber.e(e, "%s --> finalizeContactForm", this.getClass().getCanonicalName());
             }
-        }
+        }*/
```
Comment (rkareko) [inline opensrp-fp/src/main/java/org/smartregister/fp/common/interactor/ContactInteractor.java]: Should this commented out block be committed?

File: opensrp-fp/src/main/java/org/smartregister/fp/common/task/FinalizeVisitFormTask.java [hunk 2]
```diff
@@ -72,37 +70,36 @@ protected void onPreExecute() {
...
         HashMap<String, String> clientProfileDetail = PatientRepository.getClientProfileDetails(baseEntityId);
         if (clientProfileDetail == null) return null;
 
-        int gestationAge = getGestationAge(clientProfileDetail);
+        String nextContactNo = clientProfileDetail.get(DBConstantsUtils.KeyUtils.NEXT_CONTACT);
         boolean isFirst = TextUtils.equals("1", clientProfileDetail.get(DBConstantsUtils.KeyUtils.NEXT_CONTACT));
-        ContactRule contactRule = new ContactRule(gestationAge, isFirst, baseEntityId);
+        if (!isFirst) {
+//            String stillOnMethod = fpFormUtils.extractStillOnMethodValue(jsonCurrentState);
+//            isFirst = !stillOnMethod.equalsIgnoreCase("yes");
```
Comment (rkareko) [inline opensrp-fp/src/main/java/org/smartregister/fp/common/task/FinalizeVisitFormTask.java]: Are these commented out lines required?

File: opensrp-fp/src/main/java/org/smartregister/fp/common/util/FPFormUtils.java [hunk 3]
```diff
@@ -47,6 +49,24 @@ public static String obtainValue(String key, JSONArray value) throws JSONExcepti
         return result;
     }
 
+    public static String extractStillOnMethodValue(String formJsonDraft) {
+        try {
+            JSONObject jsonObject = new JSONObject(formJsonDraft);
+            JSONArray step3 = jsonObject.getJSONObject("step3").getJSONArray("fields");
```
Comment (rkareko) [inline opensrp-fp/src/main/java/org/smartregister/fp/common/util/FPFormUtils.java]: Let's move the strings `step3` and `fields` to a constants file

File: opensrp-fp/src/main/java/org/smartregister/fp/common/util/FPFormUtils.java [hunk 4]
```diff
@@ -47,6 +49,24 @@ public static String obtainValue(String key, JSONArray value) throws JSONExcepti
...
+
+            for (int i = 0; i < step3.length(); i++) {
+                if (step3.getJSONObject(i).has("key") && step3.getJSONObject(i).has("value") &&
+                        step3.getJSONObject(i).get("key").equals(STILL_ON_METHOD)) {
+                    return step3.getJSONObject(i).get("value").toString();
```
Comment (rkareko) [inline opensrp-fp/src/main/java/org/smartregister/fp/common/util/FPFormUtils.java]: Let's move the strings `key` and `value` to a constants file

File: opensrp-fp/src/main/java/org/smartregister/fp/common/util/Utils.java [hunk 5]
```diff
@@ -613,6 +619,44 @@ public static ButtonAlertStatus getButtonAlertStatus(Map<String, String> details
...
+
+    public static String getFormatedDateForFPAlertRule(String triggerDate) {
+        SimpleDateFormat format = new SimpleDateFormat("dd-MM-yyyy");
+        Date newDate = null;
+        try {
+            newDate = format.parse(triggerDate);
+        } catch (ParseException e) {
+            e.printStackTrace();
+        }
+        format = new SimpleDateFormat("yyyy-MM-dd");
```
Comment (rkareko) [inline opensrp-fp/src/main/java/org/smartregister/fp/common/util/Utils.java]: Let's move the date format strings (such as `yyyy-MM-dd` ) to a constants file

File: opensrp-fp/src/main/java/org/smartregister/fp/common/util/Utils.java [hunk 6]
```diff
@@ -613,6 +619,44 @@ public static ButtonAlertStatus getButtonAlertStatus(Map<String, String> details
...
+
+    private static String getFormatDateForVisitButton(String triggerDate) {
+        SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd");
+        Date newDate = null;
+        try {
+            newDate = format.parse(triggerDate);
+        } catch (ParseException e) {
+            e.printStackTrace();
+        }
+        format = new SimpleDateFormat("dd MMM yyyy");
```
Comment (rkareko) [inline opensrp-fp/src/main/java/org/smartregister/fp/common/util/Utils.java]: Let's move the date format strings (such as `dd MMM yyyy` ) to a constants file

File: opensrp-fp/src/main/java/org/smartregister/fp/common/util/Utils.java [hunk 7]
```diff
@@ -613,6 +619,44 @@ public static ButtonAlertStatus getButtonAlertStatus(Map<String, String> details
...
+        try {
+            newDate = format.parse(triggerDate);
+        } catch (ParseException e) {
+            e.printStackTrace();
+        }
+        format = new SimpleDateFormat("yyyy-MM-dd");
+        return format.format(newDate);
+    }
+
+    private static String getFormatDateForVisitButton(String triggerDate) {
```
Comment (rkareko) [inline opensrp-fp/src/main/java/org/smartregister/fp/common/util/Utils.java]: This method and getFormatedDateForFPAlertRule() appear to be doing the same thing, the only difference being the `triggerDate` date format and the `newDate` date format.

Is it possible to combine them into a generic function?

**Review by Rkareko — DISMISSED** [review]

**Review by Rkareko — APPROVED** [review]

**Pull request merged**